cmd/manager/manager.go

```diff
@@ -0,0 +1,196 @@
+	eventHandlers  map[string][]handler.EventHandler
+	applications   map[string]application.Application
+	eventBus       bus.EventBus
+	metricBus      bus.MetricBus
+	pluginPath     string
+	logger         *logging.Logger
+)
+
+func init() {
+	transports = map[string]transport.Transport{}
+	metricHandlers = map[string][]handler.MetricHandler{}
+	eventHandlers = map[string][]handler.EventHandler{}
+	applications = map[string]application.Application{}
+	pluginPath = "/usr/lib64/sg-core"
+}
+
+func eventHandleDecorator(data []byte, call func([]byte) (data.Event, error)) {
+	e, err := call(data)
+	if err != nil {
+		logger.Metadata(logging.Metadata{"error": err})
+		logger.Error("cannot publish event to event bus")
+		return
+	}
+	eventBus.Publish(e)
+}
+
+func eventMetricDecorator(data []byte, call func([]byte) ([]data.Metric, error)) {
+	m, err := call(data)
+	if err != nil {
+		logger.Metadata(logging.Metadata{"error": err})
+		logger.Error("cannot publish event to event bus")
+		return
+	}
+	metricBus.Publish(m)
+}
+
+//SetPluginDir set directory path containing plugin binaries
+func SetPluginDir(path string) {
+	pluginPath = path
+}
+
+//SetLogger set logger
+func SetLogger(l *logging.Logger) {
+	logger = l
+}
+
+//InitTransport load tranpsort binary and initialize with config
+func InitTransport(name string, config interface{}) error {
+	n, err := initPlugin(name)
+	if err != nil {
+		return errors.Wrap(err, "failed initializing transport")
+	}
+
+	new, ok := n.(func(*logging.Logger) transport.Transport)
+	if !ok {
+		return fmt.Errorf("plugin %s constructor 'New' did not return type 'transport.Transport'", name)
+	}
+
+	transports[name] = new(logger)
+
+	c, err := yaml.Marshal(config)
+	if err != nil {
+		return errors.Wrapf(err, "failed parsing transport config for '%s'", name)
+	}
+
+	err = transports[name].Config(c)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//InitApplication initialize application plugin with configuration
+func InitApplication(name string, config interface{}) error {
+	n, err := initPlugin(name)
+	if err != nil {
+		return errors.Wrap(err, "failed initializing application plugin")
+	}
+
+	new, ok := n.(func(*logging.Logger) application.Application)
+	if !ok {
+		return fmt.Errorf("plugin %s constructor 'New' did not return type 'application.Application'", name)
+	}
+
+	applications[name] = new(logger)
+
+	c, err := yaml.Marshal(config)
+	if err != nil {
+		return errors.Wrapf(err, "failed parsing application plugin config for '%s'", name)
+	}
+
+	err = applications[name].Config(c)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+//SetTransportHandlers load handlers binaries for transport
+func SetTransportHandlers(name string, handlerNames []string) error {
+	for _, hName := range handlerNames {
+		n, err := initPlugin(hName)
+		if err != nil {
+			return errors.Wrap(err, "failed initializing handler")
+		}
+
+		var hType string
+		switch new := n.(type) {
```

Comment:
Overriding builtin function new is never a good idea IMO :).

Reply:
I was thinking back to my c++ days when overloading functions was acceptable :). The reason I did that here was just so that the configuration file would not need to separate handlers into separate sections for metrics and events